File: Mbs.Api/Models/Trading/Instruments/Etf.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Runtime.Serialization;
using Mbs.Trading.Currencies;

namespace Mbs.Api.Models.Trading.Instruments
{
    /// <summary>
    /// An additional information for Exchange-Traded Funds.
    /// </summary>
    [DataContract]
    public class Etf : IValidatableObject
    {
        /// <summary>
        /// Gets or sets a currency code.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public CurrencyCode Currency { get; set; } = CurrencyCode.Xxx;

        /// <summary>
        /// Gets or sets a trading mode.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string TradingMode { get; set; }

        /// <summary>
        /// Gets or sets an ISO 10962 Classification of Financial Instruments code.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string Cfi { get; set; }

        /// <summary>
        /// Gets or sets a dividend frequency.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string DividendFrequency { get; set; }

        /// <summary>
        /// Gets or sets an exposition type.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string ExpositionType { get; set; }

        /// <summary>
        /// Gets or sets a fraction.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string Fraction { get; set; }

        /// <summary>
        /// Gets or sets a total expence ratio percentage.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string TotalExpenseRatio { get; set; }

        /// <summary>
        /// Gets or sets an index family.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string IndexFamily { get; set; }

        /// <summary>
        /// Gets or sets a launch date.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string LaunchDate { get; set; }

        /// <summary>
        /// Gets or sets an issuer.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string Issuer { get; set; }

        /// <summary>
        /// Gets or sets an Indicative Net Asset Value instrument reference.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public InstrumentReference Inav { get; set; }

        /// <summary>
        /// Gets or sets an underlying instrument reference.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public InstrumentReference Underlying { get; set; }

        /// <inheritdoc />
        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            yield break;
        }
    }
}

File: Mbs.Api.Host.Ng/ClientApp/src/app/shared/mbs/instruments/types/etf.ts
import { CurrencyCode } from '../../currencies/currency-code.enum';
import { InstrumentReference } from './instrument-reference';

/** An additional information for Exchange-Traded Funds. */
export class Etf {
  /** A currency code. */
  currency!: CurrencyCode;

  /** A trading mode. */
  tradingMode?: string;

  /** An ISO 10962 *Classification of Financial Instruments* code. */
  cfi?: string;

  /** A dividend frequency. */
  dividendFrequency?: string;

  /** An exposition type. */
  expositionType?: string;

  /** A fraction. */
  fraction?: string;

  /** A total expence ratio percentage. */
  totalExpenseRatio?: string;

  /** An index family. */
  indexFamily?: string;

  /** A launch date. */
  launchDate?: string;

  /** An issuer. */
  issuer?: string;

  /** An *Indicative Net Asset Value* instrument reference. */
  inav?: InstrumentReference;

  /** Gets or sets an underlying instrument reference. */
  underlying?: InstrumentReference;

  constructor(data?: Etf) {
    if (data) {
      for (const property in data) {
        if (data.hasOwnProperty(property)) {
          (this as any)[property] = (data as any)[property];
        }
      }
      this.inav = data.inav && !(data.inav as any).toJSON ? new InstrumentReference(data.inav) : (this.inav as InstrumentReference);
      this.underlying = data.underlying && !(data.underlying as any).toJSON ? new InstrumentReference(data.underlying) :
        (this.underlying as InstrumentReference);
    }
  }

  toJSON(data?: any) {
    data = typeof data === 'object' ? data : {};
    // tslint:disable:no-string-literal
    data['Currency'] = this.currency;
    data['TradingMode'] = this.tradingMode;
    data['Cfi'] = this.cfi;
    data['DividendFrequency'] = this.dividendFrequency;
    data['ExpositionType'] = this.expositionType;
    data['Fraction'] = this.fraction;
    data['TotalExpenseRatio'] = this.totalExpenseRatio;
    data['IndexFamily'] = this.indexFamily;
    data['LaunchDate'] = this.launchDate;
    data['Issuer'] = this.issuer;
    data['Inav'] = this.inav ? this.inav.toJSON() : (undefined as any);
    data['Underlying'] = this.underlying ? this.underlying.toJSON() : (undefined as any);
    // tslint:enable:no-string-literal
    return data;
  }
}
